### PR TITLE
Add UpdateInstantiateConfig gov proposal

### LINKED ISSUE #780: Freeze Code ID Proposal
With permissioned CosmWasm chains, it would be nice to freeze code ids from being instantiated. 

Example:  There is a security bug in a code id and a new one is deployed but you want to prevent the old one from being instantiated but at the same time keep previously instantiated contracts to keep running.

### LINKED ISSUE #810: Remove stray "sudo" in comment
In

```
  // CodeID references the new WASM codesudo
  uint64 code_id = 5 [ (gogoproto.customname) = "CodeID" ];
```

there is an extra sudo. I tried to remove it but I don't have the proto tooling to run the proto scripts. Also see `buf protoc` does not exist anymore (https://github.com/bufbuild/buf/pull/915).

### PR DESCRIPTION
**Pull request by jhernandezb**
closes #780 
closes #810

- [x] Server Side
- [x] Client Side
- [x] Testing

### CODE REVIEW

**Comment by jhernandezb** [comment]
@alpe  Regarding single update vs multiple updates on a single proposal. Should it be one proposal  at a time? or udpating multiple is fine? also different configs? Currently I have multiple configs with the same AccessConfig.

Also I was unable to run `make proto-all` only `make proto-gen` worked the other commands are giving me errors of proto definitions not found. (`lint, proto-breaking`)

**Comment by alpe** [comment]
@jhernandezb good start! You are more familiar with the concrete use case but from my perspective multiple code_ids with same AccessConfig make a lot of sense. This can be explained to the voters but also gives flexibility to address multiple code ids.

**Comment by jhernandezb** [comment]
Yes my reasoning is for example: Freeze Code ID 1,2,3,4 then you only need to apply the same config to all of them. If someone wanted different access configs then a different proposal should be made explaining the reasoning.

Will complete this PR then and push updates

**Comment by ethanfrey** [comment]
This will close #780 right?

If so, please add that to the description

**Comment by ethanfrey** [comment]
> @alpe Regarding single update vs multiple updates on a single proposal. Should it be one proposal at a time? or udpating multiple is fine? also different configs? Currently I have multiple code ids with the same AccessConfig.

I would either:
* One code Id, one instantiate config
* A list of N repeated (code_id, instantiate_config) submessages

I have no strong opinion either way. Although I guess the second would be better to match the use case you had in mind to allow multiple code_ids

**Comment by ethanfrey** [comment]
If we can get this ready by (early?) next week, I would love to get it in the upcoming v0.27 release

**Comment by jhernandezb** [comment]
working on this today can have a PR ready for review tomorrow so I think its doable to have it for v0.27

**Comment by jhernandezb** [comment]
@ethanfrey @alpe this is ready for review will be doing some manual testing.

Some notes:
- Added a new method to the keeper `SetAccessConfig`
- added testing to proposal_integration tests with different cases but let me know if it can be improved
- Moved to `A list of N repeated (code_id, instantiate_config) submessages` approach 
- CLI command is a bit tricky with that approach but added examples

**Review by alpe — CHANGES_REQUESTED** [review]
Excellent work. I am very happy with this PR.

Only some comments on rest that need be addressed before merging.

Too bad we have not dropped legacy rest support already.

File: x/wasm/client/rest/gov.go
Comment (jhernandezb) [reply x/wasm/client/rest/gov.go:431]: good catch it doesn't need to be here.

File: x/wasm/client/rest/gov.go
Comment (jhernandezb) [reply x/wasm/client/rest/gov.go:429]: this follows the protobuf defitinion 
```
repeated CodeAccessConfigUpdate code_updates = 3
```

should we rename the proto field to code_access_config_updates ? 

I think it would make sense because is just doing an specific action and not performing a full update to a code id.

thoughts?

File: x/wasm/types/exported_keepers.go
Comment (jhernandezb) [reply x/wasm/types/exported_keepers.go:58]: `NewWasmProposalHandlerX` takes `ContractOpsKeeper` as argument, not really sure how have access to this method without exporting it (not really sure if a private method makes sense on an interface)

File: x/wasm/keeper/proposal_integration_test.go
Comment (jhernandezb) [reply x/wasm/keeper/proposal_integration_test.go:770]: refactored to reuse that type.

File: x/wasm/client/rest/gov.go
Comment (jhernandezb) [reply x/wasm/client/rest/gov.go:452]: fixed

**Comment by jhernandezb** [comment]
Made some updates just have couple questions
https://github.com/CosmWasm/wasmd/pull/796#discussion_r866928010
https://github.com/CosmWasm/wasmd/pull/796#discussion_r866932343

**Comment by ethanfrey** [comment]
@alpe does this address all your concerns?

I will do a review and merge this evening / tomorrow unless there is a blocker

File: proto/cosmwasm/wasm/v1/proposal.proto
Comment (ethanfrey) [inline proto/cosmwasm/wasm/v1/proposal.proto:158]: Odd that this is 4. Not wrong, just surprising, I assume this was for some backwards compatibility with another struct?

It would be good to label it if so, with something like "2 and 3 are reserved" or such

File: x/wasm/client/cli/gov_tx.go
Comment (ethanfrey) [inline x/wasm/client/cli/gov_tx.go:722]: I'm a bit confused why this flag is here?

Usage is `tx gov submit-proposal update-instantiate-config 1,nobody 2,everybody 3,%s1l2rsakp388kuv9k8qzq6lrm9taddae7fpx59wm` right?

File: x/wasm/client/rest/gov.go
Comment (ethanfrey) [reply x/wasm/client/rest/gov.go:429]: I think `access_config_updates` or `code_access_config_updates` would work (and ideally update both in the Protobuf file as well)

File: x/wasm/keeper/proposal_integration_test.go
Comment (ethanfrey) [inline x/wasm/keeper/proposal_integration_test.go:753]: nice

File: x/wasm/types/exported_keepers.go
Comment (ethanfrey) [reply x/wasm/types/exported_keepers.go:58]: good point here. If there is a simple way to keep it private and expose to the proposal handler, maybe @alpe can make a quick PR to show how. Otherwise, I would be fine to merge like this.

**Review by ethanfrey — COMMENTED** [review]
lgtm. I would love to see a manual test (as there are all kinds of codecs and middleware called in the real proposal stack), but good stuff.

I added a few comments on naming, which would be nice polish.
A CHANGELOG entry (one line pointing to this PR) would also be nice.

Happy to approve once those are done (just mention me in a comment and I will double check)

File: proto/cosmwasm/wasm/v1/proposal.proto
Comment (jhernandezb) [reply proto/cosmwasm/wasm/v1/proposal.proto:158]: it should be 2 will fix.

**Comment by jhernandezb** [comment]
@ethanfrey addressed your comments and also here is a demo
https://www.loom.com/share/e99721a095d4428d80f81496daaee2ef
(ignore the print long on submit proposal, I had that as debug on my local cosmos-sdk dependency)

File: x/wasm/client/cli/gov_tx.go
Comment (ethanfrey) [reply x/wasm/client/cli/gov_tx.go:722]: Good point, can you make a follow up (cleanup issue), so someone can look into that later.
Not blocking merge at all, but don't want to forget it

**Review by ethanfrey — APPROVED** [review]
Great stuff and loving the loom video!

Thank you very much

**Pull request merged**
